### PR TITLE
Feature/select

### PR DESCRIPTION
**Pull request by sebastiansandqvist**
This is not ready to be merged. Publishing here a bit early so that I can get some feedback before going on to implement the `<MediaSelect>`.

<details>
  <summary>Example code to demonstrate the API (basically demo/src/index.jsx)</summary>

```jsx

const selectOpts = [
  { label: 'Option #1', uniqueId: 'option-id-1' },
  { label: 'Option #2', uniqueId: 'option-id-2' },
  { label: 'Option #3', uniqueId: 'option-id-3' },
];

const selectOptsWithDescription = [
  { description: 'Hello World', label: 'Option #1', uniqueId: 'option-id-1' },
  { description: 'Hello Quartz', label: 'Option #2', uniqueId: 'option-id-2' },
  { description: 'Hello everyone', label: 'Option #3', uniqueId: 'option-id-3' },
];

<Select options={selectOpts} />
<Select options={selectOpts} inline />
<Select options={selectOpts} color='gray' inline />
<Select options={selectOpts} color='white' inline />
<Select options={selectOpts} color='teal' inline />
<Select options={selectOpts} dropdownPosition='above' />
<Select options={selectOpts} triggerPlaceholder='Custom label' inline />
<Select options={selectOpts} dropdownPosition='above' caretAlign='left' inline />
<Select options={selectOpts} dropdownPosition='above' caretAlign='center' inline />
<Select options={selectOpts} dropdownPosition='above' caretAlign='right' inline />
<Select options={selectOpts} caretAlign='left' inline />
<Select options={selectOpts} caretAlign='center' inline />
<Select options={selectOpts} caretAlign='right' inline />
<Select options={selectOpts} inline Trigger={({ isOpen, onOpenToggle }) => <button onClick={() => onOpenToggle(!isOpen)}>Choose something ({isOpen ? 'close' : 'open'})</button>} />
<Select options={selectOpts} multiSelect />
<Select options={selectOpts} search={(query) => {}} />
<Select options={selectOpts} multiSelect search={(query) => {}} />
<Select options={selectOpts} isLoading />
<Select options={selectOptsWithDescription} />
```
</details>

See also demo/src/demo-select.jsx for the implementation of a smart parent component wrapper for the select. It looks quite complex because of the `search`, which I want to figure out a solution to. 

<details>
  <summary>So the simplified version of it is this:</summary>

```jsx
class StatefulSelect extends Component {

  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
      selectedOptions: {},
      selectedLabel: '',
    };
    this.handleChange = this.handleChange.bind(this);
    this.setOpen = this.setOpen.bind(this);
  }

  setOpen(isOpen) {
    this.setState({ isOpen });
  }

  handleChange(selectedOptions, selectedLabel, itemToggled) {
    this.setState({ selectedOptions, selectedLabel });
  }

  render() {
    return (
      <Select
        {...this.props}
        isOpen={this.state.isOpen}
        selectedOptions={this.state.selectedOptions}
        onSelectionToggle={this.handleChange}
        onOpenToggle={this.setOpen}
        triggerLabel={this.state.selectedLabel}
      />
    );
  }
}
```
</details>

### CODE REVIEW

File: components/Select/SelectDropdownHOC.jsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          return;
+        }
+        onSelectionToggle(newSelection, label, item);
+      };
+      return <Option key={item.uniqueId} onOptionToggle={onToggle} isSelected={Boolean(selectedOptions[item.uniqueId])} {...item} />;
+    };
+
+    return (
+      <div className='c-select--dropdown bg-white border radius fill-width is-open' ref={manualOffsetV(dropdownPosition === 'above', VERTICAL_OFFSET)}>
+        <If condition={Boolean(search)}>
```
Comment (erikmoldovan) [inline components/Select/SelectDropdownHOC.jsx]: I really like that If util

File: components/Select/standard/SelectDropdownOption.jsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import Icon from '../../Icon';
+import { If } from '../../util';
+
+// TODO: this is a hack, we should have this in css if possible to make a PR to Quartz css
+// (This fixes wrapping issues in `inline` select dropdowns)
+const listStyle = { whiteSpace: 'nowrap' };
```
Comment (erikmoldovan) [inline components/Select/standard/SelectDropdownOption.jsx]: Agreed

**Comment by sebastiansandqvist** [comment]
### Ready for review!

There are many places where this component gets complex internally. I would like to make it as easy to understand and use as possible, so if there's anything I can do to help with that then please let me know.

On that note, I'd go in this order to get an understanding of how it works:
1. `npm start` and look at the browser console when using the `<select>`s
2. See the usage in `demo/src/index.jsx`
3. See the smart component in `demo/src/demo-select.jsx`
4. See the implementation in `components/Select/standard/Select.jsx`.

#### "processing" state:

I initially wanted to be able to just have something like this for option state (as we discussed):

```js
this.state = {
  optionState: {
    option1: 'checked',
    option2: 'unchecked',
    option3: 'unchecked',
    option4: 'processing',
  },
  ...
}
```

There was a lot of added complexity involved with doing it that way, so I opted for the way quartz currently handles processing state, which is to pass an array of processing option ids as a prop.

So currently it works like this:

```js
this.state = {
  selectedOptions: {
    option1: true
    option2: false
    option3: false
    option4: false
  },
  processingOptions: [ 'option4' ],
  ...
}
```

The stateful component might look like this:

```jsx
function removeFromArray(array, item) {
  const index = array.indexOf(item);
  if (index !== -1) { array.splice(index, 1); }
  return array;
}

class ProcessingSelect extends Component {
  constructor() {
    super();
    this.state = {
      isOpen: false,
      selectedOptions: {},
      selectedLabel: '',
      processingOptions: [],
    };
    this.setOpen = this.setOpen.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  setOpen(isOpen) {
    this.setState({ isOpen });
  }

  handleChange(selectedOptions, selectedLabel, item) {
    const { processingOptions } = this.state;
    this.setState({ processingOptions: processingOptions.concat(item.uniqueId) });
    setTimeout(() => {
      this.setState({
        selectedOptions,
        selectedLabel,
        processingOptions: removeFromArray(processingOptions, item.uniqueId),
      });
    }, 500);
  }

  render() {
    return (
      <Select
        {...this.props}
        isOpen={this.state.isOpen}
        selectedOptions={this.state.selectedOptions}
        onSelectionToggle={this.handleChange}
        onOpenToggle={this.setOpen}
        triggerLabel={this.state.selectedLabel}
      />
    );
  }
}

// used like this:
<ProcessingSelect options={[/* some options go here */]} />
```

Initially I didn't like this way of handling "processing" state as much, but it has a few benefits.

1. No string constants. You can do `if (this.state.selectedOptions[foo]) { ... }` instead of `if (this.state.selectedOptions[foo] === 'checked') { ... }`
2. Since most of the time you won't be using this option, the default api is simpler.
3. Simpler implementation internally (easier to refactor)

(If this isn't ok, I can go back to the first way.)

#### Still left to do:

1. Add tests
2. Maybe implement `action` and `onAction` (although I do not think this currently works, so I am not sure if it is needed). This image is [a component that uses those props](https://github.com/vhx/crystal/blob/0e8cedfbc8ad2d75a9b5313aae988de2333354ca/app/assets/javascripts/admin/content/collections/components/selectbar/selectbar.ui.js#L74), and the `action` does not show up at all. screen shot 2017-06-19 at 4 15 20 pm
3. Maybe add custom dropdown menu vertical offsets. Do you think this would be needed? I didn't see them in use in Crystal (but didn't look too hard).
4. Add `:hover` state to standard dropdowns. I wanted to implement this in quartz css, but if that is not possible for now then I can use `onMouseOver` and `onMouseOut` to do the same thing.

File: components/Select/Trigger.jsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  });
+}
+
+const Trigger = ({ color, isOpen, onOpenToggle, triggerLabel, triggerPlaceholder }) => (
+  <span className={getTriggerClass({ color })} onClick={() => onOpenToggle(!isOpen)}>{triggerLabel || triggerPlaceholder}</span>
+);
+
+Trigger.propTypes = {
+  color: PropTypes.oneOf([ 'gray', 'white', 'teal' ]).isRequired,
+  isOpen: PropTypes.bool.isRequired, // TODO: could remove this propType, but it would probably be useful to anyone specifying a custom <Trigger> element
```
Comment (sebastiansandqvist) [inline components/Select/Trigger.jsx]: I think I'll remove this `todo:` comment and keep the propType there. It's passed into the custom trigger. If you were making a custom trigger, you would likely want it to be able to know if the select `isOpen`

**Pull request merged**